src/com/google/javascript/jscomp/Compiler.java

```diff
@@ -140,6 +140,9 @@
   // The JS source inputs
   private List<CompilerInput> inputs;
 
+  // The JS source inputs
+  private Map<String, CompilerInput.ModuleType> moduleTypesByName;
```

Comment:
1. Could you make this `final`?
2. (From @blickly) I think it would be easier to follow if this map were indexed by CompilerInput rather than by String.  It's not as clear to me what the "canonical" name of a file should be, and it looks like the CompilerInput is always what is used to compute the names that are currently used.

Reply:
For 2: The main point is that the module type needs looked up for import statements - and you can't get the compiler input for that. You get a module path (and can easily get module name). See https://github.com/google/closure-compiler/pull/2729/files#diff-d1b3eef5af5a700763a2fce367ab27ddR153

Reply:
To better illustrate:

```
const foo = require('bar');
```

We need to know what type of module `'bar'` is. That's the primary purpose of this map.

Reply:
FYI I'm open to other suggestions on where to put this and how to utilize this information. I'm not 100% happy with this, but I don't have any better suggestions either.

Reply:
Ah, OK, thanks for the detailed explanation, Chad.  That makes sense. Somehow I had forgotten that the ModuleLoader is all path-based and doesn't have access to the module contents.